<launch>
  <!-- This is for running detection of nodes and links without setting all the parameters. Unstable launch file, will change often. -->
  <!-- Object file Name without extension-->
  <arg name="object"         default="link_uniform,node_uniform" doc="Object mesh names. Must be specified in the same order as in the SVM." />
  <!--<arg name="data_path"      default="$(find sp_segmenter)/data" doc="SP segmenter data folder" />-->
  <arg name="data_path"      default="$(env HOME)/data" doc="SP segmenter data folder" />
  <arg name="svm_path"       default="link_node_svm" doc="Object svm folder name located in data path" />
  <arg name="debug_args"     default="true"/>

  <!-- Segmentation Parameters parameters-->
  <arg name="compute_pose"          default="true" />
  <arg name="visualization"         default="false" />

  <arg name="objRecRANSACdetector"   default="StandardRecognize" doc="ObjeRecRANSAC detection options (case sensitive): StandardRecognize - standard ObjRecRANSAC routine.   StandardBest - Get best pose only or all pose above minConfidence.  GreedyRecognize - greedy detector finding highest confidence object then deleting those pixels." />
  <arg name="objRecRANSACdetectorInGripper"   default="StandardBest" doc="same as ObjectRecRANSACdetector, but this is triggered by the Gripper specific service callback" />
  <arg name="minConfidence"  default="0.20" doc="(float) confidence threshold for recognizing objects with ObjRecRANSAC" />
  <arg name="use_cuda"       default="true" doc="Use NVidia CUDA GPU acceleration to speed up pose estimation. When true and CUDA not available, prints warnings then runs without cuda "/>

  <arg name="use_external_segmentation" default="false" doc="This will make the sp_segmenter to just perform pose estimation with the assumption that the point cloud input's Red channel is the label of the object class."/>

  <arg name="useMedianFilter" default="false" doc="Apply median filter to point cloud input before processing it" />
  <arg name="maxFrames"       default="15" doc="Maximum frame averaged for svm segmentation "/>

  <arg name="useTableSegmentation" default="true" doc="use marker-based table segmentation at all or just handle raw point clouds. True is strongly recommended."/>
  <arg name="useCropBox" default="true" doc="use crop box based on table center."/>
  <arg name="aboveTableMin"     default="0.015" doc="The minimum point cloud distance from segmented table. Increase the value if some parts of table point cloud still remains after segmentation" />
  <arg name="aboveTableMax"     default="0.5" doc="The maximum point cloud distance from segmented table." />

  <arg name="tableTF"        default="table_frame" doc="Any TF frame located in the table can be used for table segmentation" />
  <arg name="saveTabledir"   default="$(arg data_path)" doc="Save/Load folder for segmenting table" />
  <arg name="loadTable"      default="true" doc="load the table corner position from table.pcd in data folder. If not available or set false, spServer will get segment new table corner position around tableTF"/>
  <arg name="saveTable"      default="true" doc="Save new table corner positions in data folder if the table is not loaded/available."/>
  <arg name="useBinarySVM"   default="false" doc="This will do binary classification before multi class SVM."/>
  <arg name="useMultiClassSVM"   default="true" doc="This will do multiclass classification before Object Ransac."/>

  <arg name="setObjectOrientation" default="true" doc="use specified TF as Object preferred orientation" />
  <arg name="preferredOrientation" default="world" doc="use this TF as the preferred object orientation" />
  <arg name="gripperTF"      default="gripper_center" doc="The gripper tf where target object would be attached" />
  <arg name="useObjectPersistence" default="false" doc="use Rtree to check for existing objects to preserve orientation and TF names" />
  <arg name="max_distance_persistence" default="0.025" doc="Maximum allowed distance for a pose to be considered as the previous pose" />

  <arg name="NodeName"       default="costar_segmenter" doc="The name of the ros topic" />

  <!-- SPSegmenterNode subscriber/publisher args -->
  <arg name="pcl_in"         default="/camera/depth_registered/points" />
  <arg name="pcl_out"        default="points_out"/>
  <arg name="poses_out"      default="poses_out"/>
  <arg name="useTF"          default="true" doc="Whether use TF frames instead of pose array for object pose representation" />
  <!--<node pkg="sp_segmenter" type="SPSegmenterServer" name="$(arg NodeName)" args="-p $(arg params)" launch-prefix="valgrind - -log-file=$(env HOME)/valgrind.log">  -->
  <node pkg="sp_segmenter" type="SPSegmenterServer" name="$(arg NodeName)" args="-p " output="screen"> 
    <param name="compute_pose"   type="bool"   value="$(arg compute_pose)"/>
    <param name="visualization"  type="bool"   value="$(arg visualization)"/>
    <!-- SPSegmenterNode Node arg pass -->
    <param name="cur_name"   type="str" value="$(arg object)" />
    <param name="objRecRANSACdetector" type="str" value="$(arg objRecRANSACdetector)" />
    <param name="minConfidence"  type="double" value="$(arg minConfidence)"/>
    <param name="aboveTableMin"     type="double" value="$(arg aboveTableMin)"/>
    <param name="aboveTableMax"     type="double" value="$(arg aboveTableMax)"/>
    <param name="use_cuda"   type="bool"   value="$(arg use_cuda)"/>
    <param name="useTableSegmentation" type="bool" value="$(arg useTableSegmentation)"/>
    <param name="useCropBox" type="bool" value="$(arg useCropBox)"/>
    <param name="cropBoxX" type="double" value="0.35"/>
    <param name="cropBoxY" type="double" value="0.35"/>
    <param name="cropBoxZ" type="double" value="0.35"/>
    <param name="cropBoxTableZ" type="double" value="0.02"/>

    <param name="cropBoxGripperX" type="double" value="0.05"/>
    <param name="cropBoxGripperY" type="double" value="0.125"/>
    <param name="cropBoxGripperZ" type="double" value="0.05"/>
  
    <param name="tableTF"        type="str" value="$(arg tableTF)"/>
    <param name="saveTable_directory"   type="str" value="$(arg saveTabledir)" />
    <param name="saveTable"   type="bool" value="$(arg saveTable)" />
    <param name="loadTable"   type="bool" value="$(arg loadTable)" />
    <param name="tableDistanceThreshold" type="double" value="0.02"/>
    <param name="tableAngularThreshold"  type="double" value="2"/>
    <param name="tableMinimalInliers"  type="double" value="7500"/>
    
    <param name="useBinarySVM"   type="bool" value="$(arg useBinarySVM)" />
    <param name="useMultiClassSVM"   type="bool" value="$(arg useMultiClassSVM)" />
    <param name="maxFrames"   type="int"  value="$(arg maxFrames)" />
    <param name="useMedianFilter"   type="bool"  value="$(arg useMedianFilter)" />
    <param name="use_external_segmentation" type="bool" value="$(arg use_external_segmentation)" />
    
    <param name="GripperTF"  type="str" value="$(arg gripperTF)"/>
    <param name="useObjectPersistence"   type="bool" value="$(arg useObjectPersistence)" />
    <param name="max_distance_persistence"  type="double" value="$(arg max_distance_persistence)"/>

    <param name="setObjectOrientation"   type="bool" value="$(arg setObjectOrientation)" />
    <param name="preferredOrientation" type="str" value="$(arg preferredOrientation)" />
    
    <!-- SPSegmenterNode subscriber/publisher args pass -->
    <param name="POINTS_IN"  type="str" value="$(arg pcl_in)"/>
    <param name="POINTS_OUT" type="str" value="$(arg pcl_out)"/>
    <param name="POSES_OUT"  type="str" value="$(arg poses_out)"/>
    <!-- param name="pairWidth"  type="double" value="$(arg pairWidth)"/ -->

    <param name="svm_path"   type="str" value="$(arg data_path)/$(arg svm_path)/" />
    <param name="useTF"      type="bool" value="$(arg useTF)" />

    <!-- objectDatabase -->
    <rosparam command="load" file="$(find sp_segmenter)/launch/objectDatabase.yaml" />
  </node>

  <group ns="$(arg NodeName)">
    <include file="$(find sp_segmenter)/launch/parameters.launch">
        <arg name="data_path" value ="$(arg data_path)"/>
    </include>
  </group>
</launch>
